### PR TITLE
Created a way to bulk load the RTree, with associated changes to the Lay...

### PR DESCRIPTION
**Pull request by phil20686**
...er interface to make it accessible through Layer.addAllNodes(Iterable<Node>)

Key Nodes: created a much more efficient way to load an RTree, references for the algorithms given in javaDoc comments as appropriate. Basic details below

1) if the insert large compared with the tree, rebuild the tree from scratch using a top down method which minimises overlapping.
2) Else use a seeded bulk clustering, i.e. use the existing tree structure to partition the data, and then bulk insert by building an R tree from scratch and inserting its children at appropriate points.
3) Outliers and underpacked nodes are dealt with through one by one insertion at the end. This means worst case behaviour is still polynomial if your data is constructed so that all inserts are always outliers. 
4) This algorithm builds a tree with a loading factor of 0.7, this is slightly suboptimal from a querying point of view, but enormously beneficial if you plan further inserts.

### CODE REVIEW

File: src/main/java/org/neo4j/gis/spatial/DefaultLayer.java [hunk 1]
```diff
@@ -19,10 +19,7 @@
  */
 package org.neo4j.gis.spatial;
 
-import java.util.Arrays;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.Set;
+import java.util.*;
```
Comment (ehx-v1) [inline src/main/java/org/neo4j/gis/spatial/DefaultLayer.java]: In general, it's no good idea to import a whole package. The classes `Arrays`, `HashSet`, `Iterator`, `Iterable` and `Set` are far from everything that is to `java.util`, so you import tons of useless stuff which might even get a hindrance. For example, you might want to use a different `Date` class further on.

File: src/main/java/org/neo4j/gis/spatial/rtree/RTreeIndex.java [hunk 2]
```diff
@@ -19,10 +19,15 @@
  */
 package org.neo4j.gis.spatial.rtree;
 
-import java.util.ArrayList;
-import java.util.Iterator;
-import java.util.List;
```
Comment (ehx-v1) [inline src/main/java/org/neo4j/gis/spatial/rtree/RTreeIndex.java]: Same as further up.

**Comment by ehx-v1** [comment]
Also, this is not quite well-readable. Your alignment makes the diffs completely go bananas.

File: src/main/java/org/neo4j/gis/spatial/DefaultLayer.java (hunk 1 above)
Comment (phil20686) [inline src/main/java/org/neo4j/gis/spatial/DefaultLayer.java]: Yes I am surprised that I did that. I am opposed to wildcard imports...

**Comment by phil20686** [comment]
Yes, I write it in in telliJ and it autoformats files. I didnt really realise at the time, I have to alwasy remember when I use intelliJ to make a commit which is just the reformatting. 

Incidentally, although this code works, there is a subtle flaw that makes the tree suboptimal. Somewhere in the build tree function it splits up the incoming nodes. iirc its in a function called partition which I added to the RTree file which attempts to presort the nodes, but the method for partitioning doesnt respect the sorting. This means that the R tree that is created has very high overlap. That was why this pull request was never merged into the main branch. I was going ot fix it but somehow I never got around to it.

File: src/main/java/org/neo4j/gis/spatial/DefaultLayer.java (hunk 1 above)
Comment (ehx-v1) [inline src/main/java/org/neo4j/gis/spatial/DefaultLayer.java]: One thought: Perhaps IntelliJ did that? You should check it, and perhaps find out how you can disable automatic wildcard imports.

File: src/main/java/org/neo4j/gis/spatial/rtree/RTreeIndex.java [hunk 3]
```diff
@@ -38,895 +43,1176 @@
...
+    }
+
+    public RTreeIndex(GraphDatabaseService database, Node rootNode, EnvelopeDecoder envelopeDecoder, int maxNodeReferences) {
+        this.database = database;
+        this.rootNode = rootNode;
+        this.envelopeDecoder = envelopeDecoder;
+        this.maxNodeReferences = maxNodeReferences;
+
+        if (envelopeDecoder == null) {
+            throw new NullPointerException("envelopeDecoder is NULL");
```
Comment (ehx-v1) [inline src/main/java/org/neo4j/gis/spatial/rtree/RTreeIndex.java]: Better: "Envelope decoder should not be null"

File: src/main/java/org/neo4j/gis/spatial/DefaultLayer.java [hunk 4]
```diff
@@ -83,6 +80,20 @@ public SpatialDatabaseRecord add(Node geomNode) {
         return new SpatialDatabaseRecord(this, geomNode, geometry);
     }
 
+    public void addAllNodes(Iterable<Node> geomNodes){
+        ArrayList<Node> nodes = new ArrayList<Node>();
```
Comment (jexp) [inline src/main/java/org/neo4j/gis/spatial/DefaultLayer.java]: use java8 streams instead

File: src/main/java/org/neo4j/gis/spatial/indexprovider/LayerNodeIndex.java [hunk 5]
```diff
@@ -145,6 +147,31 @@ public void add( Node geometry, String key, Object value )
 
     }
 
+    /**
+     * This will efficiently add a large number of nodes to the index at once, because it will
+     * collect all the rTree insertions and only insert them once.
+     * @param nodes
+     */
+    public void add(List<Node> nodes){
+        List<Node> nodesToBeInsertedIntoRtree = new ArrayList<Node>();
```
Comment (jexp) [inline src/main/java/org/neo4j/gis/spatial/indexprovider/LayerNodeIndex.java]: pre-size list

**Comment by craigtaverner** [comment]
This work was separately merged into the branch at https://github.com/neo4j-contrib/spatial/tree/bulk-update, which was used to make improvements, which were then also merged into the release branches. It was released as version 0.24 for Neo4j 3.0 and 3.1

**Pull request closed without merging**